File: 프로그래머스/3/64062. 징검다리 건너기/징검다리 건너기.cpp
#include <string>
#include <vector>

using namespace std;

bool check(int n, int k, vector<int> v){
    int cnt = 0;
    for(auto c : v){
        if(c - n < 0) cnt++;
        else cnt = 0;
        
        if(cnt >= k) return false;
    }
    return true;
}

int solution(vector<int> stones, int k) {
    int answer = 0;
    int Max = 0;
    for(auto stone : stones)
        Max = max(Max, stone);
    
    int start = 1;
    int end = Max;
    
    while(start<=end){
        int mid = (start + end) / 2;
        
        if(check(mid,k,stones)){
            start = mid + 1;
            answer = mid;
        }
        else
            end = mid - 1;
    }

    return answer;
}

File: 프로그래머스/3/64062. 징검다리 건너기/징검다리 건너기.java
class Solution {
    public int solution(int[] stones, int k) {
        int answer = 0;
        int n = stones.length;
        
        int max = 0;
        for(int x : stones){
            max = Math.max(x, max);
        }
        
        int start = 1;
        int end = max + 1;
        
        while(start < end){
            int mid = (start + end) / 2;
            int flag = 0;
            int cnt = 0;
            for(int i = 0; i < n; i++){
                
                if(stones[i] < mid){ //돌을 다 못밟을 때의 경우가 연속 K 까지면 실패
                    cnt++;
                    if(cnt >= k){
                        flag = 1; break;
                    }
                }else cnt = 0;
            }
            
            
            if(flag == 1){
                end = mid;
            }
            else 
                start = mid + 1;
        }
        answer = start - 1;
        return answer;
    }
}